drivers/audio/pdm/pdm_nrfx.c

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright (c) 2020 Oticon A/S
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <nrfx_pdm.h>
+#include <drivers/audio/pdm.h>
+#include <hal/nrf_gpio.h>
+#include <stdbool.h>
+
+#define LOG_LEVEL CONFIG_PDM_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(pdm_nrfx);
+
+#define DT_DRV_COMPAT nordic_nrf_pdm
+DEVICE_DECLARE(pdm_0);
+
+#define CLK_PIN       DT_PROP(DT_NODELABEL(pdm), clk_pin)
+#define DIN_PIN       DT_PROP(DT_NODELABEL(pdm), din_pin)
+
+#define PDM_STACK_SIZE 256
+#define PDM_PRIORITY   5
+
+#define PDM_NRFX_NUMBER_OF_BUFFERS 2
+#define PDM_NRFX_BUFFER_SIZE       256
+
+K_THREAD_STACK_DEFINE(pdm_stack, PDM_STACK_SIZE);
+
+struct k_work_q m_pdm_work_q;
+
+struct pdm_nrfx_config {
+	nrfx_pdm_config_t initial_config;
+};
+
+struct pdm_nrfx_data {
+	pdm_data_handler_t data_handler;
+};
+
+struct buffer_released_data {
+	struct k_work work;
+	int16_t buffer[PDM_NRFX_BUFFER_SIZE];
+};
+
+static struct pdm_nrfx_data m_pdm_nrfx_data;
+static struct buffer_released_data m_buff_released;
+static int16_t m_next_buffer[PDM_NRFX_BUFFER_SIZE][PDM_NRFX_NUMBER_OF_BUFFERS];
+static uint8_t m_active_buffer;
+
+
+static void release_buffer(struct k_work *item)
+{
+	struct buffer_released_data *data =
+		CONTAINER_OF(item, struct buffer_released_data, work);
+
+	m_pdm_nrfx_data.data_handler(data->buffer, PDM_NRFX_BUFFER_SIZE);
+}
+
+
+static int pdm_nrfx_set_data_handler(struct device *dev,
+				     pdm_data_handler_t handler_func)
+{
+	if (handler_func == NULL) {
+		LOG_ERR("Data handler cannot be NULL.\n");
+		return -EINVAL;
+	}
+	m_pdm_nrfx_data.data_handler = handler_func;
+	return 0;
+}
+
+
+static void pdm_nrfx_event_handler(nrfx_pdm_evt_t const *p_evt)
+{
+	if (p_evt->error == NRFX_PDM_ERROR_OVERFLOW) {
+		LOG_ERR("Overflow error when handling event.\n");
+		return;
+	}
+
+	/* If a buffer was requested,
+	 * provide a new one, and alternate the active buffer
+	 */
+	if (p_evt->buffer_requested) {
+		nrfx_err_t result;
+
+		result = nrfx_pdm_buffer_set(&m_next_buffer[0][m_active_buffer],
+						PDM_NRFX_BUFFER_SIZE);
+
+		if (result != NRFX_SUCCESS) {
+			LOG_ERR("Failed to set new buffer, error %d.\n",
+					result);
+		}
+		m_active_buffer = (~m_active_buffer) & 0x01;
+	}
+
+	/* If a buffer has been released,
+	 * save it and submit it to the workqueue for the event handler
+	 */
+	if (p_evt->buffer_released != NULL) {
+		uint16_t i;
+
+		for (i = 0; i < PDM_NRFX_BUFFER_SIZE; i++) {
+			m_buff_released.buffer[i] = p_evt->buffer_released[i];
```

Comment:
I doesn't seem to be a good solution to copy all those buffers. You also don't check in any way if the previous contents of `m_buff_released.buffer` has been already processed in the work queue (and in the application; is the application supposed to copy the buffer that it is provided by the data handler, or process it entirely in the handler call?).

Buffers should be rather allocated from a pool (like `k_mem_slab`) and passed via the data handler to the user application and freed there. See for instance how it is done in [I2S](https://github.com/zephyrproject-rtos/zephyr/blob/69867a48dbe1833730517730c4f76c490ae1a467/include/drivers/i2s.h#L381).

Reply:
You're very right. I have changed it to use `k_mem_slab`s in the latest patch